cmd/mungedocs/util.go

```diff
@@ -112,72 +168,118 @@ func endMungeTag(desc string) string {
 	return fmt.Sprintf("<!-- END MUNGE: %s -->", desc)
 }
 
-// Calls 'replace' for all sections of the document not in ``` / ``` blocks. So
-// that you don't have false positives inside those blocks.
-func replaceNonPreformatted(input []byte, replace func([]byte) []byte) []byte {
-	f := splitByPreformatted(input)
-	output := []byte(nil)
-	for _, block := range f {
-		if block.preformatted {
-			output = append(output, block.data...)
-		} else {
-			output = append(output, replace(block.data)...)
+type mungeLine struct {
+	data         string
+	preformatted bool
+	header       bool
+	link         bool
+	beginTag     bool
+	endTag       bool
+}
+
+type mungeLines []mungeLine
+
+func (m1 mungeLines) Equal(m2 mungeLines) bool {
+	if len(m1) != len(m2) {
+		return false
+	}
+	for i, _ := range m1 {
+		if m1[i].data != m2[i].data {
+			return false
 		}
 	}
-	return output
+	return true
 }
 
-type fileBlock struct {
-	preformatted bool
-	data         []byte
+func (mlines mungeLines) Bytes() []byte {
```

Comment:
If this is solely for writing out to a file, perhaps it would be more efficient to loop over the lines and write them one by one without constructing a large intermediate string and then copying it to a byte slice.

You could even just have this type implement the io.WriterTo interface.